### PR TITLE
Update CA2000 to exclude System.IO.StringReader

### LINKED ISSUE #3212: CA2000 - Exclude StringReader by default
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

[CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000)

#### Repro steps

```
var x = new StringReader("abc");
```

#### Expected behavior
`StringReader` is a well-known type that does not need to be disposed, so no warning.

#### Actual behavior
`CA2000` is reported.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #3212

### CODE REVIEW

File: src/Utilities/Compiler/Options/CategorizedAnalyzerConfigOptions.cs [hunk 1]
```diff
@@ -113,7 +113,7 @@ public T GetOptionValue<T>(string optionName, DiagnosticDescriptor rule, TryPars
                 return defaultValue;
             }
 
-            return (T)_computedOptionValuesMap.GetOrAdd($"{rule.Id}.{optionName}", _ => ComputeOptionValue(optionName, rule, tryParseValue, defaultValue))!;
+            return (T)_computedOptionValuesMap.GetOrAdd($"{rule.Id}.{optionName}.{defaultValue?.GetHashCode()}", _ => ComputeOptionValue(optionName, rule, tryParseValue, defaultValue))!;
```
Comment (Evangelink) [inline src/Utilities/Compiler/Options/CategorizedAnalyzerConfigOptions.cs]: I had to update the cache to allow callers with a different "forced value" to have a different result and not the same cache entry. We might want to have a constraint on `T` if we go-on with the `GetHashCode` method. Alternatively, we could change the method signature to have the default value being a `string` and not a `T` and we would call `TryParse` on this `string` entry.

Anyways let me know which direction you think is best.

**Comment by Evangelink** [comment]
Tagging @sharwell @mavasani

**Comment by Evangelink** [comment]
ping @sharwell

File: src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs [hunk 2]
```diff
@@ -247,28 +248,51 @@ static SymbolNamesWithValueOption<INamedTypeSymbol>.NameParts GetParts(string na
         private static SymbolNamesWithValueOption<TValue> GetSymbolNamesWithValueOption<TValue>(
             this AnalyzerOptions options,
             string optionName,
-            string? namePrefixOpt,
```
Comment (mavasani) [inline src/Utilities/Compiler/Options/AnalyzerOptionsExtensions.cs]: I think it is preferable to keep this as a required parameter so every new callsite has to make an explicit decision on supported name prefix.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.cs [hunk 3]
```diff
@@ -62,7 +62,8 @@ private DisposeAnalysis(DisposeAnalysisDomain analysisDomain, DisposeDataFlowOpe
                 disposeOwnershipTransferLikelyTypes, disposeOwnershipTransferAtConstructor,
                 disposeOwnershipTransferAtMethodCall, trackInstanceFields, exceptionPathsAnalysis,
                 performCopyAnalysis: analyzerOptions.GetCopyAnalysisOption(rule, defaultValue: performCopyAnalysisIfNotUserConfigured, cancellationToken),
-                excludedSymbols: analyzerOptions.GetExcludedSymbolNamesWithValueOption(rule, wellKnownTypeProvider.Compilation, cancellationToken),
+                // StringReader doesn't need to be disposed: https://docs.microsoft.com/en-us/dotnet/api/system.io.stringreader?view=netframework-4.8
+                excludedSymbols: analyzerOptions.GetExcludedSymbolNamesWithValueOption(rule, wellKnownTypeProvider.Compilation, cancellationToken, defaultForcedValue: $"T:{WellKnownTypeNames.SystemIOStringReader}"),
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.cs]: Wouldn't it be simpler to just update this existing special case for `Task` to also include this new type: https://github.com/dotnet/roslyn-analyzers/blob/81512992e784b6108a88d93df0ffe0f72ca0af5c/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.DisposeDataFlowOperationVisitor.cs#L93-L98

File: src/Utilities/Compiler/Options/CategorizedAnalyzerConfigOptions.cs (hunk 1 above)
Comment (mavasani) [reply src/Utilities/Compiler/Options/CategorizedAnalyzerConfigOptions.cs]: Yeah, this looks fragile. I'd rather avoid all changes in this PR to any of the options infrastructure and just make an isolated fix to Dispose analysis, as suggested in https://github.com/dotnet/roslyn-analyzers/pull/3299/files#r384039514

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DisposeObjectsBeforeLosingScopeTests.cs [hunk 4]
```diff
@@ -11956,5 +11956,92 @@ void M2()
...
+
+Public Class C
+    Public Sub New()
+        Dim x = new StringReader(""abc"")
+    End Sub
+End Class");
+        }
+
+        [Fact, WorkItem(3212, "https://github.com/dotnet/roslyn-analyzers/issues/3212")]
+        public async Task StringReader_CustomSymbolExclusion_NoDiagnostic()
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DisposeObjectsBeforeLosingScopeTests.cs]: I have kept this test just to ensure that there is no impact when changing the excluded_symbols. If you think that's useless, I will remove it.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs [hunk 5]
```diff
@@ -3638,6 +3639,11 @@ public sealed override TAbstractAnalysisValue VisitLocalFunction(ILocalFunctionO
         /// </summary>
         protected INamedTypeSymbol? GenericIEquatableNamedType { get; }
 
+        /// <summary>
+        /// <see cref="INamedTypeSymbol"/> for <see cref="System.IO.StringReader"/>
+        /// </summary>
+        protected INamedTypeSymbol? StringReaderType { get; }
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs]: You need to add this to the public API surface for AnalyzerUtilities package.

Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs]: `F:\workspace\_work\1\s\src\Utilities\FlowAnalysis\FlowAnalysis\Framework\DataFlow\DataFlowOperationVisitor.cs(3645,56): error RS0016: Symbol 'StringReaderType.get' is not part of the declared API.`

Comment (Evangelink) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs]: Yes sorry I missed the warning in VS :disappointed:

**Pull request merged**